File: e-dashboard-ui/src/ClientApp/src/views/Diagram/DiagramLib/Shapes/TextShape.ts
import { IPoint } from './interfaces';
import { TextStyle } from './TextStyle';
import { SVGUtil } from '../Util';
import { ITextShape, ITextStyle } from './interfaces';
import { Point } from './Shapes';

export class TextShape implements ITextShape {
    public $type: string ;
    public id:string;
    public text: string;
    public offset: IPoint;
    public location: IPoint;
    public isVisible: boolean;
    public style: ITextStyle;
    public angle: number = 0;

    constructor(id: string, text: string, offset: IPoint, angle: number = 0) {
        this.$type = "TextShape";
        this.id = id;
        this.text = text;
        this.offset = offset;
        this.location = new Point(0, 0);
        this.angle = angle;
        this.isVisible = true;
        this.style = new TextStyle();
    }

    public draw(ctx: CanvasRenderingContext2D) {
        let font: string = this.style.fontStyle + ' ' + this.style.fontSize + 'pt ' + this.style.fontFamily;
        ctx.save();
        ctx.globalAlpha = this.style.opacity;
        ctx.lineWidth = this.style.strokeWidth;
        ctx.fillStyle = this.style.fillStyle;
        ctx.font = font;
        ctx.textAlign = this.style.textAlign;
        ctx.textBaseline = this.style.textBaseline;
        ctx.translate(this.location.x + this.offset.x, this.location.y + this.offset.y);
        ctx.rotate(this.angle * Math.PI / 180);
        ctx.fillText(this.text, 0, 0);
        ctx.restore();
    }

    public move(to: IPoint) {
        this.location = to;
    }

    public rotate(angle: number): void {
        this.angle = angle;
    }

    public inTextZone(mouse: IPoint): boolean {
        var ret: boolean = false;
        return ret;
    }

    public toSVG(): string {
        let x = this.location.x + this.offset.x;
        let y = this.location.y + this.offset.y;

        var markup: Array<string> = [];
        markup.push(
            '\t\t<text ', this.getSvgId(),
            'x="' + x + '" ',
            'y="' + y + '" ',
            'font-family="' + this.style.fontFamily.replace(/"/g, '\'') + '" ',
            'font-size="' + this.style.fontSize + '" ',
            'font-style="' + this.style.fontStyle + '" ',
            'alignment-baseline="' + this.style.textBaseline + '" ',
            'text-anchor="' + this.style.textAnchor + '" ',
            'style="', SVGUtil.getSvgStyles(this.style), '"', ' >\n',
            '\t\t\t', this.text, '\n',
            '\t\t</text>\n'
        );

        return markup.join('');
    }

    public wrapText(ctx: CanvasRenderingContext2D, text: string, x: number, y: number, maxWidth: number, lineHeight: number) {
        var words = text.split(' ');
        var line = '';

        for (var n = 0; n < words.length; n++) {
            var testLine = line + words[n] + ' ';
            var metrics = ctx.measureText(testLine);
            var testWidth = metrics.width;
            if (testWidth > maxWidth && n > 0) {
                ctx.fillText(line, x, y);
                line = words[n] + ' ';
                y += lineHeight;
            }
            else {
                line = testLine;
            }
        }
        ctx.fillText(line, x, y);
    }

    protected getSvgId(): string {
        var ret = this.id ? 'id="' + this.id + '" ' : ' ';
        return ret;
    }
}

File: e-dashboard-api/src/Domain/Diagram/Shapes/TextShape.cs

using System.Text.Json.Serialization;

namespace EnergyDashboard.Domain.Diagram.Shapes;

[JsonDerivedType(typeof(TextShape), nameof(TextShape))]
public class TextShape
{
    public TextShape()
    {
        this.Location = new Point(0, 0);
        this.IsVisible = true;
        this.Style = new TextStyle();
        this.Angle = 0.0;
    }
    public TextShape(string id, string text, Point offset) : this()
    {
        this.Id = id;
        this.Text = text;
        this.Offset = offset;
    }


    public string Id { get; set; }
    public string Text { get; set; }
    public Point Offset { get; set; }
    public Point Location { get; set; }
    public bool IsVisible { get; set; }
    public TextStyle Style { get; set; }
    public double Angle { get; set; }

    public string toSVG()
    {
        var x = this.Location.X + this.Offset.X;
        var y = this.Location.Y + this.Offset.Y;

        string svg =
            "\t\t<text id=\"" + this.Id + "\" " +
                    "x=\"" + x + "\" " +
                    "y=\"" + y + "\" " +
                    "font-family=\"" + this.Style.FontFamily + "\" " +
                    "font-size=\"" + this.Style.FontSize + "\" " +
                    "font-style=\"" + this.Style.FontStyle + "\" " +
                    "alignment-baseline=\"" + this.Style.TextBaseline + "\" " +
                    "text-anchor=\"" + this.Style.TextAnchor + "\" " +
                    "style=\"" + Util.Utilility.GetSvgStyles(this.Style) + "\"" + " >\n" +
                    "\t\t\t" + this.Text + "\n" +
                    "\t\t</text>\n";

        return svg;
    }
}
